src/engine/search_engine_data.cpp

```diff
@@ -93,25 +110,97 @@ void SearchEngineData::InitializeOrClearManyToManyThreadLocalStorage(const unsig
     }
 }
 
-void SearchEngineData::InitializeOrClearMultiLayerDijkstraThreadLocalStorage(
-    const unsigned number_of_nodes)
+// CH
+using CH = routing_algorithms::ch::Algorithm;
+template SearchEngineData<CH>::SearchEngineHeapPtr SearchEngineData<CH>::forward_heap_1;
+template SearchEngineData<CH>::SearchEngineHeapPtr SearchEngineData<CH>::reverse_heap_1;
+template SearchEngineData<CH>::SearchEngineHeapPtr SearchEngineData<CH>::forward_heap_2;
+template SearchEngineData<CH>::SearchEngineHeapPtr SearchEngineData<CH>::reverse_heap_2;
+template SearchEngineData<CH>::SearchEngineHeapPtr SearchEngineData<CH>::forward_heap_3;
+template SearchEngineData<CH>::SearchEngineHeapPtr SearchEngineData<CH>::reverse_heap_3;
+template SearchEngineData<CH>::ManyToManyHeapPtr SearchEngineData<CH>::many_to_many_heap;
+
+template void
+SearchEngineData<routing_algorithms::ch::Algorithm>::InitializeOrClearFirstThreadLocalStorage(
+    unsigned number_of_nodes);
+
+template void
+SearchEngineData<CH>::InitializeOrClearSecondThreadLocalStorage(unsigned number_of_nodes);
+
+template void
+SearchEngineData<CH>::InitializeOrClearThirdThreadLocalStorage(unsigned number_of_nodes);
+
+template void
+SearchEngineData<CH>::InitializeOrClearManyToManyThreadLocalStorage(unsigned number_of_nodes);
+
+// CoreCH
+using CoreCH = routing_algorithms::corech::Algorithm;
+template SearchEngineData<CoreCH>::SearchEngineHeapPtr SearchEngineData<CoreCH>::forward_heap_1;
+template SearchEngineData<CoreCH>::SearchEngineHeapPtr SearchEngineData<CoreCH>::reverse_heap_1;
+template SearchEngineData<CoreCH>::SearchEngineHeapPtr SearchEngineData<CoreCH>::forward_heap_2;
+template SearchEngineData<CoreCH>::SearchEngineHeapPtr SearchEngineData<CoreCH>::reverse_heap_2;
+template SearchEngineData<CoreCH>::SearchEngineHeapPtr SearchEngineData<CoreCH>::forward_heap_3;
+template SearchEngineData<CoreCH>::SearchEngineHeapPtr SearchEngineData<CoreCH>::reverse_heap_3;
+template SearchEngineData<CoreCH>::ManyToManyHeapPtr SearchEngineData<CoreCH>::many_to_many_heap;
+
+template void
+SearchEngineData<CoreCH>::InitializeOrClearFirstThreadLocalStorage(unsigned number_of_nodes);
```

Comment:
Hrm we might want to simplify the interface here a little bit. There is no reason to ever get a reference to one of the heaps without clearing it first. So what about:

```
std::array<SearchEngineHeapPtr, MAX_HEAPS> heaps;

SearchHeap& GetHeap(HeapID id, unsigned number_of_nodes) {
     // init or clear code
     return *heaps[id];
}
```

That would also enable us to make all `{forward,backward}_heap_{1,2,3}` private and hide the thread-specific pointer. Also there is no reason why we have a distinction between forward and backward heaps.